src/mem/remoteallocator.h

```diff
@@ -22,34 +24,39 @@ namespace snmalloc
       AtomicCapPtr<Remote, CBAlloc> next{nullptr};
     };
 
-    /*
-     * We embed the size class in the bottom 8 bits of an allocator ID (i.e.,
-     * the address of an Alloc's remote_alloc's message_queue; in practice we
-     * only need 7 bits, but using 8 is conjectured to be faster).  The hashing
-     * algorithm of the Alloc's RemoteCache already ignores the bottom
-     * "initial_shift" bits, which is, in practice, well above 8.  There's a
-     * static_assert() over there that helps ensure this stays true.
-     *
-     * This does mean that we might have message_queues that always collide in
-     * the hash algorithm, if they're within "initial_shift" of each other. Such
-     * pairings will substantially decrease performance and so we prohibit them
-     * and use SNMALLOC_ASSERT to verify that they do not exist in debug builds.
+    /**
+     * Cache the size class of the object to improve performance.
      */
-    alloc_id_t alloc_id_and_sizeclass;
+    sizeclass_t sizeclasscache;
 
+    /**
+     * Set up a remote object.  Potentially cache sizeclass and allocator id.
+     *
+     * Current implementation does not cache the allocator id due to security concerns.
+     * Alternative implementations may store a hash of the allocator id, so that amplification costs can be mitigated
+     * on CHERI with MTE.
+     */
     void set_info(alloc_id_t id, sizeclass_t sc)
     {
-      alloc_id_and_sizeclass = (id & ~SIZECLASS_MASK) | sc;
+      UNUSED(id);
+      sizeclasscache = sc;
     }
 
-    alloc_id_t trunc_target_id()
+    /**
+     * Return allocator for this object.  This may perform amplification.
+     */
+    template<typename LargeAlloc>
+    static alloc_id_t trunc_target_id(CapPtr<Remote, CBAlloc> r, LargeAlloc* large_allocator)
     {
-      return alloc_id_and_sizeclass & ~SIZECLASS_MASK;
+      // Rederive allocator id.
```

Comment:
This might be expensive on CHERI, where `capptr_amplify` has to do some maths and a load.  The reason for the `alloc_id_and_sizeclass` cache was so that we could get to the "Merely routing" branch without needing to `capptr_amplify`.  If it turns out that that is having real performance impact on non-CHERI, then probably we should either have the AAL manage the contents of `Remote`s or we should, here, switch on `aal_supports<StrictProvenance>` (or some other condition?).

Reply:
I tried to leave the relevant arguments around so that either could be done for this reason.  Yeah, just having an actual capability/pointer to the allocator seems a big security risk, so having an encoded version/hash or something would be better.

Reply:
I suppose I'm content to have this land and I'll investigate the performance impact on my CHERIfication branch.

Reply:
It now keeps both implementations, if an ifdef for which to use.